### DISCUSSION TITLE
Is it possible to parametrize on the values of a fixture? How/Why not?

### DISCUSSION THREAD

**Discussion by kunaltyagi** [pytest-dev/pytest#11359] [Q&A] [answered]
I have a fixture, which returns a configuration dict. I want to parametrize my test on the values of a particular key. Eg: 
```py
@pytest.fixture(scope="session")
def config():
  return {"greet": ["Hola", "Hello", "Hi"],  # just a sample, actually generated from config files
          "count": [1,2,3,4,5,6,7,8,9],
          "cities": ["SF", "NY"],
          }

#####
@pytest.mark.parametrize("greet", HOW_TO_GET_THE_LIST_HERE)
def test_single_greeting(greet):  # want just the greetings here, one by one, no count/cities
    # checks if the greetings fit some criteria
```
I can perform a loop inside a normal test, but this approach as a few drawbacks such as:
* hiding the details when a test fails (for which input?)
* linear instead of parallel execution

What doesn't work:
* Indirection using `lazy_fixture` (not iterable)
* Loop + yield in a different fixture (can have just 1 yield)
* Separate function to load, which is called by the `config` fixture slows down the tests
    ```py
    def get_config(key=None): return expensive_load_config()[key] if key else expensive_load_config()
    @pytest.fixture(scope="session")
    def config(): return get_config()
    ```

---

Ideal usage: something similar to the following:
```py
@pytest.mark.parametrize("greeting": config()["greet"])
def test_single_greeting(greeting):
    pass
```

```py
def test_single_greeting(config, subtests):
    for greeting in config["greet"]:
        with subtests.test(msg=f"test_with_{greeting}")
            pass
```

**Comment by RonnyPfannschmidt** [comment]
parameterization currently happens before fixtures get executed

currently i believe some major internal changes are needed to enable that

**Reply by kunaltyagi** [reply]
Is there any other way to register tests in a way that fixtures would be called as usual? If there's a method to register tests inside a test, that would also work for my use-case since the mothership test would appear as a normal test to pytest

**Reply by RonnyPfannschmidt** [reply, marked as answer]
The subtests plugin provides that feature

**Reply by kunaltyagi** [reply]
That solved this. Thanks a ton
